ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader.cpp

```diff
@@ -68,20 +71,36 @@ Datadog::Uploader::upload(ddog_prof_Profile& profile)
         return ret;
     }
 
-    // Build the request object
-    const ddog_prof_Exporter_File file = {
-        .name = to_slice("auto.pprof"),
-        .file = ddog_Vec_U8_as_slice(&encoded->buffer),
-    };
-    auto build_res = ddog_prof_Exporter_Request_build(ddog_exporter.get(),
-                                                      encoded->start,
-                                                      encoded->end,
-                                                      ddog_prof_Exporter_Slice_File_empty(),
-                                                      { .ptr = &file, .len = 1 },
-                                                      nullptr,
-                                                      encoded->endpoints_stats,
-                                                      nullptr,
-                                                      nullptr);
+    std::vector<ddog_prof_Exporter_File> files_to_send = { {
+      .name = to_slice("auto.pprof"),
+      .file = ddog_Vec_U8_as_slice(&encoded->buffer),
+    } };
+
+    std::string json_str;
+
+    // DEV: re-consider this is ok to do so, as calling is_enabled/serialize will acquire
```

Comment:
Yeah, good point.  The check for enablement is non-atomic.

I guess the only way to solve this is actually internal to the CodeProvenance object itself.  You could have a `try_serialize_to_json_str()` or whatever that internally checks for enablement and returns an error or whatever if it isn't enabled.

In practice, I guess the worst-case scenario is that we lose some code provenance data under very weird circumstances?  So not a super big deal to solve.

Reply:
Oh, I was referring to the fact that this function is run with profile lock held and we're also going to take the lock in CodeProvenance too.

For your comment, `serialize_to_json_str()` already checks for `enabled` so code is simplified here.